server/src/main/java/org/elasticsearch/search/profile/SearchProfileResults.java

```diff
@@ -79,9 +90,24 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         for (String key : sortedKeys) {
             builder.startObject();
             builder.field(ID_FIELD, key);
-            shardResults.get(key).toXContent(builder, params);
+
+            ShardProfileId shardProfileId = parseProfileShardId(key);
```

Comment:
Hi Ben,

> I think this is good for backwards compatibility, but what do you think of actually serializing those objects back instead of encoding them in the id and parsing them out again? It seems brittle to me to store important info in a plain text string that is assumed to be formatted in a specific way.

In terms of brittleness, I marked this PR as partially addressing the original issue: https://github.com/elastic/elasticsearch/issues/25896, where the request was basically to split the composite ID and maybe add some additional information. They recommended not changing the ID itself because that would be a backwards incompatible change.

So the brittleness is already baked in when that composite ID was created and now we need to support that composite ID in order to not make breaking changes.

In my [analysis comment on that issue](https://github.com/elastic/elasticsearch/issues/25896#issuecomment-1565087817), I analyzed all the formats of that ID and how it's generated, indicating that it would be fairly straightforward to parse that out. Adding new info is problematic though because that composite ID is used as a Map key (https://github.com/elastic/elasticsearch/pull/96396/files#diff-582e9063bae3a8d070700f30a160c2647fd69bff489591ce5a16821bac63af6aR52) and that Map is passed around quite a bit through multiple layers before it gets to the place where it is rendered as XContent. IIRC (it's been two weeks since I looked at this), the likely place to serialize the data is in [SearchShardTarget](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/SearchShardTarget.java), but that doesn't make out to the place the XContent is rendered (only the composite ID map key).

I think we'd need to refactor a fair bit of code to avoid using that ID as the map key. So I decided to try to go the simple path first - just parsing out the existing info from that map key.

We could certainly try the bigger refactoring if we think it's worth it and want additional things not in the composite id, like `cluster_uuid`.


> the brittleness could be acceptable if test coverage was adequate.

In terms of better test coverage, the approach I took was bottom up from the possible formats that are generated by [SearchShardTarget.toString()](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/SearchShardTarget.java#L117), which is the composite ID map key.

I can enhance [the test I added](https://github.com/elastic/elasticsearch/pull/96396/files#diff-14544e5bfd8cc0411e14ddfff362b9862f32eeb25879f94ada3d7ed6f8f8fadaR35) to use the SearchShardTarget.toString() method to generate the input.

I'm open to suggestions on how else to test it. Since the node_id and possibly other fields would be randomized in yamlRestTests. The test would need to parse out the composite on the fly and I'm not sure how I'd do that in a yamlRestTest. Do you have ideas on an additional good way to test this?